### DISCUSSION TITLE
Axes values sticking when switching modes

### DISCUSSION THREAD

**Discussion by Bettarecognize** [WhiteMagic/JoystickGremlin#574] [Q&A]
Hello,

Firstly,  thank you for this awesome program.  I've been loving using it.

My problem is the program freezes the input of the axes in the default mode when switching to Mode 1 or Mode 2 until I return to the default mode and input is registered on those X, Y, Z axes again.  Same thing for the X Rotation, Y Rotation, Z rotation values when switching from Mode 1 or Mode 2 back to default.

I'm trying to setup each of my joysticks to double their axes in game.  I have 2 Vjoys setup with 7 axes each.  So the left stick my default mode uses Vjoy 1 and the X, Y, Z axes and the 'Mode 1' uses the Vjoy1 X Rotation, Y Rotation and Z Rotation axes but keeps the original X, Y, Z axes for the right stick on Vjoy2. The right stick is setup similarly so that the default mode is Vjoy2 X, Y, Z axes and 'Mode 2' X Rotation, Z Rotation axes but keeps the original X, Y, Z axes for the left stick on Vjoy1.

To switch modes I used a temporary switch mode basic that is activated to Mode 1 or Mode 2 when joystick button 1 is pressed on the left or right stick respectively and then another temporary switch mode basic that is activated to default when joystick button one is released.

Is there a better way to set this up so the Axes values will clear when switching modes?

Thank you

**Comment by WhiteMagic** [comment]
As Gremlin reacts to inputs there is no reason for it to update an axis that is not mapped. From your description it's not fully clear what behaviour you really want and the solution is different depending on that.

If you want to reset axes to a specific value when you switch a mode you can do that with a Macro. However, if you want to mirror some physical axis value then you have to use a plugin as you don't get access to that information within a Macro. An example of a plugin that does something slightly in that direction is this one here https://gist.github.com/WhiteMagic/f6dab456dc0cc3b7496b768ab735f4ba

**Reply by Bettarecognize** [reply]
Maybe I can explain my situation and desired outcome more clearly.  I'm playing Farming Simulator 25 with dual sticks and using the left stick axes for driving, turning and the right stick for camera pan, tilt, zoom on the default mode.  When I goto Modified mode 1 I want the left stick to control the same driving and turning, but the right stick to control a different action like fork lift boom up/down and tilt forward/backward.  The only issue I have is that when I switch mode, if there is a value for the vjoy axis like x axis is forward at 75% it will hold that value while mode switched.  So what happens is that I will enter mode 1 to boom up on the forklift and then return to default mode but the boom will continue moving on it's own while I'm moving the camera.

In this picture, I was making movements on Axis 1, 2, 3 while in default mode and then switch to Modified 2 mode to control Axis 4, 5, 6 but it froze the input values for Axis 1, 2, 3 

![image](https://github.com/user-attachments/assets/1ff4cb20-569e-4f72-94bf-894605c6bc15)


I think I need to find a method to reset the value of the Vjoy axis when changing modes but that plugin script only refreshes the joystick axis.

I'm trying to see if I can figure out how to do this in a Macro, but I think this might be over my head a bit.

**Reply by WhiteMagic** [reply]
The solution will likely need two parts.

**Part 1:**
To keep the left stick working the same in both modes you need to use the mode inheritance. If _mode 1_ is a child of _default_ then it will automatically do the same thing as the parent mode unless you define another action. So this would mean that the left stick would not define remaps and would keep controlling the motion of the vehicle.

**Part 2:**
As the right stick should change it's behaviour you want to have different remaps on that one. Now depending on what the resetting behaviour should look like you need a simple (Macro) or complex (Plugin) solution. If all you want is to set the _default_ remap axis of the right stick to a specific value, all you need is the Macro action. If you want to do some special memory/resetting etc. logic then you'll need a plugin.

**Reply by Bettarecognize** [reply]
So I do have the axes remapped differently between default mode and mode 1/mode 2.  Mode 1 and Mode 2 are both childs of default mode.

I'm having some success with macros activated by the same button as the temporary mode switch.  The macro will activate and clear the axes when the button is pressed it clears Axis 1, 2, 3.  The second macro is set to activate when the button is released and clear Axis 4, 5, 6 but it doesn't seem to be triggering on the button release.

![image](https://github.com/user-attachments/assets/fc8c44e8-1c04-469a-83b5-14d89c3d5ce5)

![image](https://github.com/user-attachments/assets/b197da9b-b05d-4c39-a1c0-984260f9faf5)

Is there a better way to set this up?  It would be really nice if I could set a macro to run on every mode switch.

Thank you

**Reply by WhiteMagic** [reply]
If you need a Macro to trigger on button release you'll have to add a condition for that, as by default the Macro will only trigger on button press. As for triggering actions on mode change etc., no there are no callbacks other than input value changes in Gremlin.
